// datocms-client/lib/contentfulImport/createFields.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports["default"] = void 0;

var _ora = _interopRequireDefault(require("ora"));

var _progress = _interopRequireDefault(require("./progress"));

var _toApiKey = require("./toApiKey");

var _datoFieldTypeFor = _interopRequireDefault(require("./datoFieldTypeFor"));

var _datoLinkItemTypeFor = _interopRequireDefault(require("./datoLinkItemTypeFor"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { "default": obj }; }

function _createForOfIteratorHelper(o, allowArrayLike) { var it; if (typeof Symbol === "undefined" || o[Symbol.iterator] == null) { if (Array.isArray(o) || (it = _unsupportedIterableToArray(o)) || allowArrayLike && o && typeof o.length === "number") { if (it) o = it; var i = 0; var F = function F() {}; return { s: F, n: function n() { if (i >= o.length) return { done: true }; return { done: false, value: o[i++] }; }, e: function e(_e) { throw _e; }, f: F }; } throw new TypeError("Invalid attempt to iterate non-iterable instance.\nIn order to be iterable, non-array objects must have a [Symbol.iterator]() method."); } var normalCompletion = true, didErr = false, err; return { s: function s() { it = o[Symbol.iterator](); }, n: function n() { var step = it.next(); normalCompletion = step.done; return step; }, e: function e(_e2) { didErr = true; err = _e2; }, f: function f() { try { if (!normalCompletion && it["return"] != null) it["return"](); } finally { if (didErr) throw err; } } }; }

function _unsupportedIterableToArray(o, minLen) { if (!o) return; if (typeof o === "string") return _arrayLikeToArray(o, minLen); var n = Object.prototype.toString.call(o).slice(8, -1); if (n === "Object" && o.constructor) n = o.constructor.name; if (n === "Map" || n === "Set") return Array.from(o); if (n === "Arguments" || /^(?:Ui|I)nt(?:8|16|32)(?:Clamped)?Array$/.test(n)) return _arrayLikeToArray(o, minLen); }

function _arrayLikeToArray(arr, len) { if (len == null || len > arr.length) len = arr.length; for (var i = 0, arr2 = new Array(len); i < len; i++) { arr2[i] = arr[i]; } return arr2; }

function asyncGeneratorStep(gen, resolve, reject, _next, _throw, key, arg) { try { var info = gen[key](arg); var value = info.value; } catch (error) { reject(error); return; } if (info.done) { resolve(value); } else { Promise.resolve(value).then(_next, _throw); } }

function _asyncToGenerator(fn) { return function () { var self = this, args = arguments; return new Promise(function (resolve, reject) { var gen = fn.apply(self, args); function _next(value) { asyncGeneratorStep(gen, resolve, reject, _next, _throw, "next", value); } function _throw(err) { asyncGeneratorStep(gen, resolve, reject, _next, _throw, "throw", err); } _next(undefined); }); }; }

var _default = /*#__PURE__*/function () {
  var _ref2 = _asyncToGenerator( /*#__PURE__*/regeneratorRuntime.mark(function _callee(_ref) {
    var itemTypes, datoClient, contentfulData, spinner, contentTypes, fieldSize, progress, fieldsMapping, _iterator, _step, _loop;

    return regeneratorRuntime.wrap(function _callee$(_context2) {
      while (1) {
        switch (_context2.prev = _context2.next) {
          case 0:
            itemTypes = _ref.itemTypes, datoClient = _ref.datoClient, contentfulData = _ref.contentfulData;
            spinner = (0, _ora["default"])('').start();
            _context2.prev = 2;
            contentTypes = contentfulData.contentTypes;
            fieldSize = contentTypes.map(function (contentType) {
              return contentType.fields.length;
            }).reduce(function (acc, length) {
              return acc + length;
            }, 0);
            progress = new _progress["default"](fieldSize, 'Creating fields');
            spinner.text = progress.tick();
            fieldsMapping = {};
            _iterator = _createForOfIteratorHelper(contentTypes);
            _context2.prev = 9;
            _loop = /*#__PURE__*/regeneratorRuntime.mark(function _loop() {
              var contentType, contentTypeApiKey, itemType, _iterator2, _step2, contentfulField, position, validators, fieldAttributes, datoField;

              return regeneratorRuntime.wrap(function _loop$(_context) {
                while (1) {
                  switch (_context.prev = _context.next) {
                    case 0:
                      contentType = _step.value;
                      contentTypeApiKey = (0, _toApiKey.toItemApiKey)(contentType.sys.id);
                      fieldsMapping[contentTypeApiKey] = [];
                      itemType = itemTypes.find(function (iT) {
                        return iT.apiKey === contentTypeApiKey;
                      });
                      _iterator2 = _createForOfIteratorHelper(contentType.fields);
                      _context.prev = 5;

                      _iterator2.s();

                    case 7:
                      if ((_step2 = _iterator2.n()).done) {
                        _context.next = 22;
                        break;
                      }

                      contentfulField = _step2.value;
                      position = contentType.fields.indexOf(contentfulField);
                      validators = {};

                      if (contentfulField.type === 'Link' && contentfulField.linkType === 'Entry') {
                        validators = {
                          itemItemType: {
                            itemTypes: (0, _datoLinkItemTypeFor["default"])({
                              itemTypes: itemTypes,
                              field: contentfulField
                            })
                          }
                        };
                      }

                      if (contentfulField.type === 'Array' && contentfulField.items.type === 'Link' && contentfulField.items.linkType === 'Entry') {
                        validators = {
                          itemsItemType: {
                            itemTypes: (0, _datoLinkItemTypeFor["default"])({
                              itemTypes: itemTypes,
                              field: contentfulField.items
                            })
                          }
                        };
                      }

                      fieldAttributes = {
                        label: contentfulField.name,
                        fieldType: (0, _datoFieldTypeFor["default"])(contentfulField),
                        localized: contentfulField.localized,
                        apiKey: (0, _toApiKey.toFieldApiKey)(contentfulField.id),
                        position: position,
                        validators: validators
                      };

                      if (contentfulField.id === contentType.displayField && contentfulField.type === 'Symbol') {
                        fieldAttributes.appearance = {
                          editor: 'single_line',
                          parameters: {
                            heading: true
                          },
                          addons: []
                        };
                      }

                      _context.next = 17;
                      return datoClient.fields.create(itemType.id, fieldAttributes);

                    case 17:
                      datoField = _context.sent;
                      spinner.text = progress.tick();
                      fieldsMapping[contentTypeApiKey].push(datoField);

                    case 20:
                      _context.next = 7;
                      break;

                    case 22:
                      _context.next = 27;
                      break;

                    case 24:
                      _context.prev = 24;
                      _context.t0 = _context["catch"](5);

                      _iterator2.e(_context.t0);

                    case 27:
                      _context.prev = 27;

                      _iterator2.f();

                      return _context.finish(27);

                    case 30:
                    case "end":
                      return _context.stop();
                  }
                }
              }, _loop, null, [[5, 24, 27, 30]]);
            });

            _iterator.s();

          case 12:
            if ((_step = _iterator.n()).done) {
              _context2.next = 16;
              break;
            }

            return _context2.delegateYield(_loop(), "t0", 14);

          case 14:
            _context2.next = 12;
            break;

          case 16:
            _context2.next = 21;
            break;

          case 18:
            _context2.prev = 18;
            _context2.t1 = _context2["catch"](9);

            _iterator.e(_context2.t1);

          case 21:
            _context2.prev = 21;

            _iterator.f();

            return _context2.finish(21);

          case 24:
            spinner.succeed();
            return _context2.abrupt("return", fieldsMapping);

          case 28:
            _context2.prev = 28;
            _context2.t2 = _context2["catch"](2);
            spinner.fail();
            throw _context2.t2;

          case 32:
          case "end":
            return _context2.stop();
        }
      }
    }, _callee, null, [[2, 28], [9, 18, 21, 24]]);
  }));

  return function (_x) {
    return _ref2.apply(this, arguments);
  };
}();

exports["default"] = _default;
