// datocms-client/lib/contentfulImport/createRecords.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports["default"] = void 0;

var _ora = _interopRequireDefault(require("ora"));

var _promiseLimit = _interopRequireDefault(require("promise-limit"));

var _progress = _interopRequireDefault(require("./progress"));

var _toApiKey = require("./toApiKey");

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { "default": obj }; }

function _typeof(obj) { "@babel/helpers - typeof"; if (typeof Symbol === "function" && typeof Symbol.iterator === "symbol") { _typeof = function _typeof(obj) { return typeof obj; }; } else { _typeof = function _typeof(obj) { return obj && typeof Symbol === "function" && obj.constructor === Symbol && obj !== Symbol.prototype ? "symbol" : typeof obj; }; } return _typeof(obj); }

function ownKeys(object, enumerableOnly) { var keys = Object.keys(object); if (Object.getOwnPropertySymbols) { var symbols = Object.getOwnPropertySymbols(object); if (enumerableOnly) symbols = symbols.filter(function (sym) { return Object.getOwnPropertyDescriptor(object, sym).enumerable; }); keys.push.apply(keys, symbols); } return keys; }

function _objectSpread(target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i] != null ? arguments[i] : {}; if (i % 2) { ownKeys(Object(source), true).forEach(function (key) { _defineProperty(target, key, source[key]); }); } else if (Object.getOwnPropertyDescriptors) { Object.defineProperties(target, Object.getOwnPropertyDescriptors(source)); } else { ownKeys(Object(source)).forEach(function (key) { Object.defineProperty(target, key, Object.getOwnPropertyDescriptor(source, key)); }); } } return target; }

function _slicedToArray(arr, i) { return _arrayWithHoles(arr) || _iterableToArrayLimit(arr, i) || _unsupportedIterableToArray(arr, i) || _nonIterableRest(); }

function _nonIterableRest() { throw new TypeError("Invalid attempt to destructure non-iterable instance.\nIn order to be iterable, non-array objects must have a [Symbol.iterator]() method."); }

function _iterableToArrayLimit(arr, i) { if (typeof Symbol === "undefined" || !(Symbol.iterator in Object(arr))) return; var _arr = []; var _n = true; var _d = false; var _e = undefined; try { for (var _i = arr[Symbol.iterator](), _s; !(_n = (_s = _i.next()).done); _n = true) { _arr.push(_s.value); if (i && _arr.length === i) break; } } catch (err) { _d = true; _e = err; } finally { try { if (!_n && _i["return"] != null) _i["return"](); } finally { if (_d) throw _e; } } return _arr; }

function _arrayWithHoles(arr) { if (Array.isArray(arr)) return arr; }

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

function _createForOfIteratorHelper(o, allowArrayLike) { var it; if (typeof Symbol === "undefined" || o[Symbol.iterator] == null) { if (Array.isArray(o) || (it = _unsupportedIterableToArray(o)) || allowArrayLike && o && typeof o.length === "number") { if (it) o = it; var i = 0; var F = function F() {}; return { s: F, n: function n() { if (i >= o.length) return { done: true }; return { done: false, value: o[i++] }; }, e: function e(_e2) { throw _e2; }, f: F }; } throw new TypeError("Invalid attempt to iterate non-iterable instance.\nIn order to be iterable, non-array objects must have a [Symbol.iterator]() method."); } var normalCompletion = true, didErr = false, err; return { s: function s() { it = o[Symbol.iterator](); }, n: function n() { var step = it.next(); normalCompletion = step.done; return step; }, e: function e(_e3) { didErr = true; err = _e3; }, f: function f() { try { if (!normalCompletion && it["return"] != null) it["return"](); } finally { if (didErr) throw err; } } }; }

function _unsupportedIterableToArray(o, minLen) { if (!o) return; if (typeof o === "string") return _arrayLikeToArray(o, minLen); var n = Object.prototype.toString.call(o).slice(8, -1); if (n === "Object" && o.constructor) n = o.constructor.name; if (n === "Map" || n === "Set") return Array.from(o); if (n === "Arguments" || /^(?:Ui|I)nt(?:8|16|32)(?:Clamped)?Array$/.test(n)) return _arrayLikeToArray(o, minLen); }

function _arrayLikeToArray(arr, len) { if (len == null || len > arr.length) len = arr.length; for (var i = 0, arr2 = new Array(len); i < len; i++) { arr2[i] = arr[i]; } return arr2; }

function asyncGeneratorStep(gen, resolve, reject, _next, _throw, key, arg) { try { var info = gen[key](arg); var value = info.value; } catch (error) { reject(error); return; } if (info.done) { resolve(value); } else { Promise.resolve(value).then(_next, _throw); } }

function _asyncToGenerator(fn) { return function () { var self = this, args = arguments; return new Promise(function (resolve, reject) { var gen = fn.apply(self, args); function _next(value) { asyncGeneratorStep(gen, resolve, reject, _next, _throw, "next", value); } function _throw(err) { asyncGeneratorStep(gen, resolve, reject, _next, _throw, "throw", err); } _next(undefined); }); }; }

var _require = require('humps'),
    camelize = _require.camelize;

var _default = /*#__PURE__*/function () {
  var _ref2 = _asyncToGenerator( /*#__PURE__*/regeneratorRuntime.mark(function _callee3(_ref) {
    var itemTypes, fieldsMapping, datoClient, contentfulData, spinner, _ret;

    return regeneratorRuntime.wrap(function _callee3$(_context3) {
      while (1) {
        switch (_context3.prev = _context3.next) {
          case 0:
            itemTypes = _ref.itemTypes, fieldsMapping = _ref.fieldsMapping, datoClient = _ref.datoClient, contentfulData = _ref.contentfulData;
            spinner = (0, _ora["default"])('').start();
            _context3.prev = 2;
            return _context3.delegateYield( /*#__PURE__*/regeneratorRuntime.mark(function _callee2() {
              var entries, progress, contentfulRecordMap, recordsToPublish, limit, jobs, _iterator, _step, _loop;

              return regeneratorRuntime.wrap(function _callee2$(_context2) {
                while (1) {
                  switch (_context2.prev = _context2.next) {
                    case 0:
                      entries = contentfulData.entries;
                      progress = new _progress["default"](entries.length, 'Creating records');
                      contentfulRecordMap = {};
                      recordsToPublish = [];
                      spinner.text = progress.tick();
                      limit = (0, _promiseLimit["default"])(5);
                      jobs = [];
                      _iterator = _createForOfIteratorHelper(entries);

                      try {
                        _loop = function _loop() {
                          var entry = _step.value;
                          var contentType = entry.sys.contentType;
                          var contentTypeApiKey = (0, _toApiKey.toItemApiKey)(contentType.sys.id);
                          var itemType = itemTypes.find(function (iT) {
                            return iT.apiKey === contentTypeApiKey;
                          });
                          var itemTypeFields = fieldsMapping[contentTypeApiKey];

                          if (itemType) {
                            var emptyFieldValues = itemTypeFields.reduce(function (accFields, field) {
                              if (field.localized) {
                                var value = contentfulData.locales.map(function (locale) {
                                  return locale;
                                }).reduce(function (accLocales, locale) {
                                  return Object.assign(accLocales, _defineProperty({}, locale, null));
                                }, {});
                                return Object.assign(accFields, _defineProperty({}, camelize(field.apiKey), value));
                              }

                              return Object.assign(accFields, _defineProperty({}, camelize(field.apiKey), null));
                            }, {});
                            var recordAttributes = Object.entries(entry.fields).reduce(function (acc, _ref3) {
                              var _ref4 = _slicedToArray(_ref3, 2),
                                  option = _ref4[0],
                                  value = _ref4[1];

                              var apiKey = (0, _toApiKey.toFieldApiKey)(option);
                              var field = itemTypeFields.find(function (f) {
                                return f.apiKey === apiKey;
                              });

                              switch (field.fieldType) {
                                case 'link':
                                case 'links':
                                case 'file':
                                case 'gallery':
                                  return acc;

                                default:
                                  break;
                              }

                              if (field.localized) {
                                var localizedValue = Object.keys(value).reduce(function (innerAcc, locale) {
                                  var innerValue = value[locale];

                                  if (field.fieldType === 'lat_lon') {
                                    innerValue = {
                                      latitude: innerValue.lat,
                                      longitude: innerValue.lon
                                    };
                                  }

                                  if (field.fieldType === 'string' && Array.isArray(innerValue)) {
                                    innerValue = innerValue.join(', ');
                                  }

                                  if (field.fieldType === 'json') {
                                    innerValue = JSON.stringify(innerValue, null, 2);
                                  }

                                  return Object.assign(innerAcc, _defineProperty({}, locale, innerValue));
                                }, {});
                                var fallbackValues = contentfulData.locales.reduce(function (accLocales, locale) {
                                  return Object.assign(accLocales, _defineProperty({}, locale, localizedValue[contentfulData.defaultLocale]));
                                }, {});
                                return Object.assign(acc, _defineProperty({}, camelize(apiKey), _objectSpread(_objectSpread({}, fallbackValues), localizedValue)));
                              }

                              var innerValue = value[contentfulData.defaultLocale];

                              if (field.fieldType === 'lat_lon') {
                                innerValue = {
                                  latitude: innerValue.lat,
                                  longitude: innerValue.lon
                                };
                              }

                              if (field.fieldType === 'string' && Array.isArray(innerValue)) {
                                innerValue = innerValue.join(', ');
                              }

                              if (field.fieldType === 'json') {
                                innerValue = JSON.stringify(innerValue, null, 2);
                              }

                              return Object.assign(acc, _defineProperty({}, camelize(apiKey), innerValue));
                            }, emptyFieldValues);
                            jobs.push(limit( /*#__PURE__*/_asyncToGenerator( /*#__PURE__*/regeneratorRuntime.mark(function _callee() {
                              var record;
                              return regeneratorRuntime.wrap(function _callee$(_context) {
                                while (1) {
                                  switch (_context.prev = _context.next) {
                                    case 0:
                                      _context.next = 2;
                                      return datoClient.items.create(_objectSpread(_objectSpread({}, recordAttributes), {}, {
                                        itemType: itemType.id.toString()
                                      }));

                                    case 2:
                                      record = _context.sent;

                                      if (entry.sys.publishedVersion) {
                                        recordsToPublish.push(record.id);
                                      }

                                      spinner.text = progress.tick();
                                      contentfulRecordMap[entry.sys.id] = record.id;

                                    case 6:
                                    case "end":
                                      return _context.stop();
                                  }
                                }
                              }, _callee);
                            }))));
                          }
                        };

                        for (_iterator.s(); !(_step = _iterator.n()).done;) {
                          _loop();
                        }
                      } catch (err) {
                        _iterator.e(err);
                      } finally {
                        _iterator.f();
                      }

                      _context2.next = 11;
                      return Promise.all(jobs);

                    case 11:
                      spinner.succeed();
                      return _context2.abrupt("return", {
                        v: {
                          contentfulRecordMap: contentfulRecordMap,
                          recordsToPublish: recordsToPublish
                        }
                      });

                    case 13:
                    case "end":
                      return _context2.stop();
                  }
                }
              }, _callee2);
            })(), "t0", 4);

          case 4:
            _ret = _context3.t0;

            if (!(_typeof(_ret) === "object")) {
              _context3.next = 7;
              break;
            }

            return _context3.abrupt("return", _ret.v);

          case 7:
            _context3.next = 13;
            break;

          case 9:
            _context3.prev = 9;
            _context3.t1 = _context3["catch"](2);
            spinner.fail();
            throw _context3.t1;

          case 13:
          case "end":
            return _context3.stop();
        }
      }
    }, _callee3, null, [[2, 9]]);
  }));

  return function (_x) {
    return _ref2.apply(this, arguments);
  };
}();

exports["default"] = _default;
